framework/src/meshgenerators/RefineSidesetGenerator.C

```diff
@@ -0,0 +1,117 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RefineSidesetGenerator.h"
+#include "MooseMeshUtils.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/mesh_refinement.h"
+#include "CastUniquePointer.h"
+
+registerMooseObject("MooseApp", RefineSidesetGenerator);
+
+InputParameters
+RefineSidesetGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addClassDescription("Mesh generator which refines one or more sidesets");
+  params.addRequiredParam<MeshGeneratorName>("input", "Input mesh to modify");
+  params.addRequiredParam<std::vector<BoundaryName>>("boundaries", "The list of boundaries to be modified");
+  params.addRequiredParam<std::vector<int>>(
+      "refinement",
+      "The amount of times to refine each sideset, corresponding to their index in 'boundaries'");
+  params.addParam<bool>(
+      "enable_neighbor_refinement",
+      true,
+      "Toggles whether neighboring level one elements should be refined or not. Defaults to true");
+  MultiMooseEnum boundary_side("primary secondary both", "both");
+  params.addParam<MultiMooseEnum>("boundary_side", boundary_side, "  ");
+
+  return params;
+}
+
+RefineSidesetGenerator::RefineSidesetGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _boundaries(getParam<std::vector<BoundaryName>>("boundaries")),
+    _refinement(getParam<std::vector<int>>("refinement")),
+    _enable_neighbor_refinement(getParam<bool>("enable_neighbor_refinement")),
+    _boundary_side(getParam<MultiMooseEnum>("boundary_side"))
+{
+}
+std::unique_ptr<MeshBase>
+RefineSidesetGenerator::generate()
+{
+
+  // Get the list of boundary ids from the boundary names
+  const auto boundary_ids =
+      MooseMeshUtils::getBoundaryIDs(*_input, getParam<std::vector<BoundaryName>>("boundaries"), true);
+
+  // Check that the boundary ids/names exist in the mesh
+
+  for (std::size_t i = 0; i < boundary_ids.size(); ++i)
+    if (boundary_ids[i] == Moose::INVALID_BOUNDARY_ID)
+    {
+      if (isParamValid("_boundaries"))
+        paramError("boundaries",
+                   "The boundary '",
+                   getParam<std::vector<BoundaryName>>("boundaries")[i],
+                   "' was not found within the mesh");
+    }
+  
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  int max = *std::max_element(_refinement.begin(), _refinement.end());
+
+  
+  return recursive_refine(boundary_ids, mesh, _refinement, max);
+}
+
+std::unique_ptr<MeshBase>
+RefineSidesetGenerator::recursive_refine(std::vector<boundary_id_type> boundary_ids,
+                                    std::unique_ptr<MeshBase> & mesh,
+                                    std::vector<int> refinement,
+                                    int max,
+                                    int ref_step)
+{
+
+  if (ref_step == max)
+    return dynamic_pointer_cast<MeshBase>(mesh);
```

Comment:
why do you need to cast?

Reply:
It's what the other ones all used, I simply copied the code.